src/log.rs

```diff
@@ -0,0 +1,396 @@
+        match self.value {
+            LogValue::String(ref s) => {
+                let s: *const libc::c_char = s.to_glib_none().0;
+                glib_sys::GLogField {
+                    key: self.key.to_glib_none().0,
+                    value: s as _,
+                    length: -1,
+                }
+            }
+            LogValue::Value(ref v) => glib_sys::GLogField {
+                key: self.key.to_glib_none().0,
+                value: v.as_ptr() as _,
+                length: v.len() as _,
+            },
+        }
+    }
+}
+
+#[doc(hidden)]
+impl<'a> ToGlibPtr<'a, *const glib_sys::GLogField> for &'a [&LogField] {
+    type Storage = Vec<glib_sys::GLogField>;
+
+    #[inline]
+    fn to_glib_none(&'a self) -> Stash<'a, *const glib_sys::GLogField, Self> {
+        let fields = self.iter().map(|x| x.to_glib()).collect::<Vec<_>>();
+        Stash(fields.as_ptr() as *const _, fields)
+    }
+
+    #[inline]
+    fn to_glib_full(&self) -> *const glib_sys::GLogField {
+        let ptr = unsafe {
+            glib_sys::g_malloc0(::std::mem::size_of::<glib_sys::GLogField>() * self.len())
+                as *mut glib_sys::GLogField
+        };
+        for (pos, x) in self.iter().enumerate() {
+            unsafe {
+                *(ptr.offset(pos as _)) = x.to_glib();
+            }
+        }
+        ptr as *const _
+    }
+}
+
+#[cfg(any(feature = "v2_46", feature = "dox"))]
+pub fn log_set_handler_full<P: Fn(&str, &LogLevelFlags, &str) + Send + Sync + 'static>(
+    log_domain: Option<&str>,
+    log_levels: LogLevelFlags,
+    log_func: P,
+) -> u32 {
+    let log_func_data: Box_<P> = Box_::new(log_func);
+    unsafe extern "C" fn log_func_func<
+        P: Fn(&str, &LogLevelFlags, &str) + Send + Sync + 'static,
+    >(
+        log_domain: *const libc::c_char,
+        log_level: glib_sys::GLogLevelFlags,
+        message: *const libc::c_char,
+        user_data: glib_sys::gpointer,
+    ) {
+        let log_domain: GString = from_glib_borrow(log_domain);
+        let log_level = from_glib(log_level);
+        let message: GString = from_glib_borrow(message);
+        let callback: &P = &*(user_data as *mut _);
+        (*callback)(log_domain.as_str(), &log_level, message.as_str());
+    }
+    let log_func = Some(log_func_func::<P> as _);
+    unsafe extern "C" fn destroy_func<P: Fn(&str, &LogLevelFlags, &str) + Send + Sync + 'static>(
+        data: glib_sys::gpointer,
+    ) {
+        let _callback: Box_<P> = Box_::from_raw(data as *mut _);
+    }
+    let destroy_call4 = Some(destroy_func::<P> as _);
+    let super_callback0: Box_<P> = log_func_data;
+    unsafe {
+        glib_sys::g_log_set_handler_full(
+            log_domain.to_glib_none().0,
+            log_levels.to_glib(),
+            log_func,
+            Box_::into_raw(super_callback0) as *mut _,
+            destroy_call4,
+        )
+    }
+}
+
+pub fn log_set_handler<P: Fn(&str, &LogLevelFlags, &str) + Send + Sync + 'static>(
```

Comment:
This one will leak the callback, so remove it :) That's why the full variant exists